















<!doctype html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta http-equiv="last-modified" content="2020-12-06 21:20:29 +0000">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <!-- meta "search-domain" used for google site search function google_search() -->
    <meta name="search-domain" value="">
    <link rel="stylesheet" type="text/css" href="../assets/css/bootstrap.css" />
    <link rel="stylesheet" type="text/css" href="../assets/css/bootstrap-theme.css" />
    <link rel="stylesheet" type="text/css" href="../assets/css/lesson.css" />
    <link rel="stylesheet" type="text/css" href="../assets/css/syntax.css" />
     <link rel="stylesheet" type="text/css" href="../assets/css/fonts.css" />
     <link rel="stylesheet" type="text/css" href="/assets/css/links.css" />
    <link rel="license" href="#license-info" />

    



    <!-- Favicons for everyone -->
    <link rel="apple-touch-icon-precomposed" sizes="57x57" href="../assets/favicons/dc/apple-touch-icon-57x57.png" />
    <link rel="apple-touch-icon-precomposed" sizes="114x114" href="../assets/favicons/dc/apple-touch-icon-114x114.png" />
    <link rel="apple-touch-icon-precomposed" sizes="72x72" href="../assets/favicons/dc/apple-touch-icon-72x72.png" />
    <link rel="apple-touch-icon-precomposed" sizes="144x144" href="../assets/favicons/dc/apple-touch-icon-144x144.png" />
    <link rel="apple-touch-icon-precomposed" sizes="60x60" href="../assets/favicons/dc/apple-touch-icon-60x60.png" />
    <link rel="apple-touch-icon-precomposed" sizes="120x120" href="../assets/favicons/dc/apple-touch-icon-120x120.png" />
    <link rel="apple-touch-icon-precomposed" sizes="76x76" href="../assets/favicons/dc/apple-touch-icon-76x76.png" />
    <link rel="apple-touch-icon-precomposed" sizes="152x152" href="../assets/favicons/dc/apple-touch-icon-152x152.png" />
    <link rel="icon" type="image/png" href="../assets/favicons/dc/favicon-196x196.png" sizes="196x196" />
    <link rel="icon" type="image/png" href="../assets/favicons/dc/favicon-96x96.png" sizes="96x96" />
    <link rel="icon" type="image/png" href="../assets/favicons/dc/favicon-32x32.png" sizes="32x32" />
    <link rel="icon" type="image/png" href="../assets/favicons/dc/favicon-16x16.png" sizes="16x16" />
    <link rel="icon" type="image/png" href="../assets/favicons/dc/favicon-128.png" sizes="128x128" />
    <meta name="application-name" content="Data Carpentry - Intro to R and RStudio for Genomics"/>
    <meta name="msapplication-TileColor" content="#FFFFFF" />
    <meta name="msapplication-TileImage" content="../assets/favicons/dc/mstile-144x144.png" />
    <meta name="msapplication-square70x70logo" content="../assets/favicons/dc/mstile-70x70.png" />
    <meta name="msapplication-square150x150logo" content="../assets/favicons/dc/mstile-150x150.png" />
    <meta name="msapplication-wide310x150logo" content="../assets/favicons/dc/mstile-310x150.png" />
    <meta name="msapplication-square310x310logo" content="../assets/favicons/dc/mstile-310x310.png" />


    <!-- HTML5 shim and Respond.js for IE8 support of HTML5 elements and media queries -->
    <!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
    <!--[if lt IE 9]>
	<script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
	<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
	<![endif]-->
  <title>
  Aggregating and Analyzing Data with dplyr &ndash; Intro to R and RStudio for Genomics
  </title>

  </head>
  <body>
    


<div class="panel panel-default life-cycle">
  <div id="life-cycle" class="panel-body alpha">
    This lesson is in the early stages of development (Alpha version)
  </div>
</div>





    <div class="container">
      
















  
  










<nav class="navbar navbar-default">
  <div class="container-fluid">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>

      
      
      <a href="http://datacarpentry.org" class="pull-left">
        <img class="navbar-logo" src="../assets/img/dc-icon-black.svg" alt="Data Carpentry logo" />
      </a>
      

      
      <a class="navbar-brand" href="../index.html">Home</a>

    </div>
    <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
      <ul class="nav navbar-nav">

	
        <li><a href="../CODE_OF_CONDUCT.html">Code of Conduct</a></li>

        
	
        <li><a href="../setup.html">Setup</a></li>

        
        
        <li class="dropdown">
          <a href="../" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Episodes <span class="caret"></span></a>
          <ul class="dropdown-menu">
            
            
            <li><a href="../01-introduction/index.html">Introducing R and RStudio IDE</a></li>
            
            
            <li><a href="../02-r-basics/index.html">R Basics</a></li>
            
            
            <li><a href="../03-basics-factors-dataframes/index.html">R Basics continued - factors and data frames</a></li>
            
            
            <li><a href="../04-dplyr/index.html">Aggregating and Analyzing Data with dplyr</a></li>
            
            
            <li><a href="../05-data-visualization/index.html">Data Visualization with ggplot2</a></li>
            
            
            <li><a href="../XX-knitr-markdown/index.html">Producing Reports With knitr</a></li>
            
	    <li role="separator" class="divider"></li>
            <li><a href="../aio.html">All in one page (Beta)</a></li>
          </ul>
        </li>
        
	

	
	
        <li class="dropdown">
          <a href="../" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Extras <span class="caret"></span></a>
          <ul class="dropdown-menu">
            <li><a href="../reference.html">Reference</a></li>
            
            
            <li><a href="../about/index.html">About</a></li>
            
            
            <li><a href="../discuss/index.html">Discussion</a></li>
            
            
            <li><a href="../figures/index.html">Figures</a></li>
            
            
            <li><a href="../guide/index.html">Instructor Notes</a></li>
            
          </ul>
        </li>
	

	
        <li><a href="../LICENSE.html">License</a></li>
	
	
	<li><a href="/edit//_episodes_rmd/04-dplyr.Rmd" data-checker-ignore>Improve this page <span class="glyphicon glyphicon-pencil" aria-hidden="true"></span></a></li>
	
	
      </ul>
      <form class="navbar-form navbar-right" role="search" id="search" onsubmit="google_search(); return false;">
        <div class="form-group">
          <input type="text" id="google-search" placeholder="Search..." aria-label="Google site search">
        </div>
      </form>
    </div>
  </div>
</nav>

      






      


























  
  











<div class="row">
  <div class="col-xs-1">
    <h3 class="text-left">
      
      <a href="../03-basics-factors-dataframes/index.html"><span class="glyphicon glyphicon-menu-left" aria-hidden="true"></span><span class="sr-only">previous episode</span></a>
      
    </h3>
  </div>
  <div class="col-xs-10">
    
    <h3 class="maintitle"><a href="../">Intro to R and RStudio for Genomics</a></h3>
    
  </div>
  <div class="col-xs-1">
    <h3 class="text-right">
      
      <a href="../05-data-visualization/index.html"><span class="glyphicon glyphicon-menu-right" aria-hidden="true"></span><span class="sr-only">next episode</span></a>
      
    </h3>
  </div>
</div>

<article>
<div class="row">
  <div class="col-md-1">
  </div>
  <div class="col-md-10">
    <h1 class="maintitle">Aggregating and Analyzing Data with dplyr</h1>
  </div>
  <div class="col-md-1">
  </div>
</div>












<blockquote class="objectives">
  <h2>Overview</h2>

  <div class="row">
    <div class="col-md-3">
      <strong>Teaching:</strong> 60 min
      <br/>
      <strong>Exercises:</strong> 15 min
    </div>
    <div class="col-md-9">
      <strong>Questions</strong>
      <ul>
	
	<li><p>How can I manipulate dataframes without repeating myself?</p>
</li>
	
      </ul>
    </div>
  </div>

  <div class="row">
    <div class="col-md-3">
    </div>
    <div class="col-md-9">
      <strong>Objectives</strong>
      <ul>
	
	<li><p>Select certain columns in a data frame with the <strong><code class="language-plaintext highlighter-rouge">dplyr</code></strong> function <code class="language-plaintext highlighter-rouge">select</code>.</p>
</li>
	
	<li><p>Extract certain rows in a data frame according to logical (boolean) conditions with the <strong><code class="language-plaintext highlighter-rouge">dplyr</code></strong> function <code class="language-plaintext highlighter-rouge">filter</code>.</p>
</li>
	
	<li><p>Link the output of one <strong><code class="language-plaintext highlighter-rouge">dplyr</code></strong> function to the input of another function with the ‘pipe’ operator <code class="language-plaintext highlighter-rouge">%&gt;%</code>.</p>
</li>
	
	<li><p>Add new columns to a data frame that are functions of existing columns with <code class="language-plaintext highlighter-rouge">mutate</code>.</p>
</li>
	
	<li><p>Use the split-apply-combine concept for data analysis.</p>
</li>
	
	<li><p>Use <code class="language-plaintext highlighter-rouge">summarize</code>, <code class="language-plaintext highlighter-rouge">group_by</code>, and <code class="language-plaintext highlighter-rouge">count</code> to split a data frame into groups of observations, apply summary statistics for each group, and then combine the results.</p>
</li>
	
	<li><p>Employ the ‘split-apply-combine’ concept to split the data into groups, apply analysis to each group, and combine the results.</p>
</li>
	
	<li><p>Export a data frame to a .csv file.</p>
</li>
	
      </ul>
    </div>
  </div>

</blockquote>

<h1 id="data-manipulation-using-dplyr-and-tidyr-from-tidyverse-package">Data manipulation using <strong><code class="language-plaintext highlighter-rouge">dplyr</code></strong> and <strong><code class="language-plaintext highlighter-rouge">tidyr</code></strong> from <strong><code class="language-plaintext highlighter-rouge">tidyverse</code></strong> package</h1>

<p>Bracket subsetting is handy, but it can be cumbersome and difficult to read,
especially for complicated operations. Enter <strong><code class="language-plaintext highlighter-rouge">dplyr</code></strong>. <strong><code class="language-plaintext highlighter-rouge">dplyr</code></strong> is a package for
making tabular data manipulation easier. It pairs nicely with <strong><code class="language-plaintext highlighter-rouge">tidyr</code></strong> which enables you to swiftly convert between different data formats for plotting and analysis.</p>

<p>The <strong><code class="language-plaintext highlighter-rouge">tidyverse</code></strong> package is an “umbrella-package” that installs <strong><code class="language-plaintext highlighter-rouge">tidyr</code></strong>, <strong><code class="language-plaintext highlighter-rouge">dplyr</code></strong>, and several other packages useful for data analysis, such as  <strong><code class="language-plaintext highlighter-rouge">ggplot2</code></strong>, <strong><code class="language-plaintext highlighter-rouge">tibble</code></strong>, etc.</p>

<p>The <strong><code class="language-plaintext highlighter-rouge">tidyverse</code></strong> package tries to address 3 common issues that arise when
doing data analysis with some of the functions that come with R:</p>

<ol>
  <li>The results from a base R function sometimes depend on the type of data.</li>
  <li>Using R expressions in a non standard way, which can be confusing for new
learners.</li>
  <li>Hidden arguments, having default operations that new learners are not aware
of.</li>
</ol>

<p>First we have to make sure we have the <code class="language-plaintext highlighter-rouge">tidyverse</code> package (including <code class="language-plaintext highlighter-rouge">dplyr</code> 
and <code class="language-plaintext highlighter-rouge">tidyr</code>) installed.</p>

<h4 id="install-tidyverse">Install <code class="language-plaintext highlighter-rouge">tidyverse</code></h4>

<p><strong>ON NOTABLE RSTUDIO SERVER:</strong></p>

<p>If you’re using the EDINA Notable RStudio Server, <code class="language-plaintext highlighter-rouge">tidyverse</code> is already installed,
so you can skip this step. It’s worth knowing how to install a package for future 
though, so check out the instructions for locally-installed RStudio.</p>

<p><strong>ON LOCAL RSTUDIO:</strong></p>

<p>Install the package you want by using the <code class="language-plaintext highlighter-rouge">install.package()</code> function, with the 
name of the package you want inside the brackets, within quotes. 
If you get an error, make sure you’ve spelled the package name correctly.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">install.packages</span><span class="p">(</span><span class="s2">"tidyverse"</span><span class="p">)</span><span class="w"> </span><span class="c1">## install tidyverse</span><span class="w">
</span></code></pre></div></div>

<p>You might get asked to choose a CRAN mirror – this is asking you to
choose a site to download the package from. The choice doesn’t matter too much; I’d recommend choosing the RStudio mirror.</p>

<h4 id="load-tidyverse-package">Load <code class="language-plaintext highlighter-rouge">tidyverse</code> package</h4>

<p>To load the package (whether in Notable or running RStudio locally), type (without quotes this time!):</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1">## load the tidyverse packages, incl. dplyr</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p>You only need to install a package once per computer, but you need to load it
every time you open a new R session and want to use that package.</p>

<h2 id="what-are-dplyr-and-tidyr">What are <strong><code class="language-plaintext highlighter-rouge">dplyr</code></strong> and <strong><code class="language-plaintext highlighter-rouge">tidyr</code></strong>?</h2>

<p>The package <strong><code class="language-plaintext highlighter-rouge">dplyr</code></strong> provides easy tools for the most common data manipulation
tasks. It is built to work directly with data frames, with many common tasks
optimized by being written in a compiled language (C++). An additional feature is the
ability to work directly with data stored in an external database. The benefits of
doing this are that the data can be managed natively in a relational database,
queries can be conducted on that database, and only the results of the query are
returned.</p>

<p>This addresses a common problem with R in that all operations are conducted
in-memory and thus the amount of data you can work with is limited by available
memory. The database connections essentially remove that limitation in that you
can connect to a database of many hundreds of GB, conduct queries on it directly, and pull
back into R only what you need for analysis.</p>

<p>The package <strong><code class="language-plaintext highlighter-rouge">tidyr</code></strong> addresses the common problem of wanting to reshape your data for
plotting and use by different R functions. Sometimes we want data sets where we have one
row per measurement. Sometimes we want a data frame where each measurement type has its
own column, and rows are instead more aggregated groups
(e.g., a time period, an experimental unit like a plot or a batch number).
Moving back and forth between these formats is non-trivial, and <strong><code class="language-plaintext highlighter-rouge">tidyr</code></strong> gives you tools
for this and more sophisticated  data manipulation.</p>

<p>To learn more about <strong><code class="language-plaintext highlighter-rouge">dplyr</code></strong> and <strong><code class="language-plaintext highlighter-rouge">tidyr</code></strong> after the workshop, you may want to check out this
<a href="https://github.com/rstudio/cheatsheets/raw/master/data-transformation.pdf">handy data transformation with <strong><code class="language-plaintext highlighter-rouge">dplyr</code></strong> cheatsheet</a>
and this <a href="https://github.com/rstudio/cheatsheets/raw/master/data-import.pdf">one about <strong><code class="language-plaintext highlighter-rouge">tidyr</code></strong></a>.</p>

<p>As before, we’ll read in our variants data:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">variants</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read.csv</span><span class="p">(</span><span class="s2">"https://ndownloader.figshare.com/files/14632895"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p>We then want to check our variants object:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1">## inspect the data</span><span class="w">
</span><span class="n">str</span><span class="p">(</span><span class="n">variants</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1">## preview the data</span><span class="w">
</span><span class="n">view</span><span class="p">(</span><span class="n">variants</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<!-- COLUMN NAME INFO FOR VARIANTS DATA: "Explore the VCF format" section: -->
<!-- https://datacarpentry.org/wrangling-genomics/04-variant_calling/index.html -->

<p>Next, we’re going to learn some of the most common <strong><code class="language-plaintext highlighter-rouge">dplyr</code></strong> functions as well 
as using pipes to combine them:</p>

<ul>
  <li><code class="language-plaintext highlighter-rouge">select()</code>: subset columns</li>
  <li><code class="language-plaintext highlighter-rouge">filter()</code>: subset rows on conditions</li>
  <li><code class="language-plaintext highlighter-rouge">mutate()</code>: create new columns by using information from other columns</li>
  <li><code class="language-plaintext highlighter-rouge">group_by()</code> and <code class="language-plaintext highlighter-rouge">summarize()</code>: create summary statistics on grouped data</li>
  <li><code class="language-plaintext highlighter-rouge">arrange()</code>: sort results</li>
  <li><code class="language-plaintext highlighter-rouge">count()</code>: count discrete values</li>
</ul>

<h3 id="selecting-columns-and-filtering-rows">Selecting columns and filtering rows</h3>

<p>Using bracket notation, in order to pull out certain columns, we would have to 
know which index referred to that column, which took an extra step, and if we 
had re-ordered our columns or added one in the middle, the indexes in our script
would suddenly refer to the ‘wrong’ column, which might cause confusion!</p>

<p>As a refresher, here’s how we pulled out columns using the bracket notation:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># get columns: sample_id, REF, ALT, DP</span><span class="w">
</span><span class="n">variants</span><span class="p">[,</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">5</span><span class="p">,</span><span class="w"> </span><span class="m">6</span><span class="p">,</span><span class="w"> </span><span class="m">12</span><span class="p">)]</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>   sample_id      REF       ALT DP
1 SRR2584863        T         G  4
2 SRR2584863        G         T  6
3 SRR2584863        G         T 10
4 SRR2584863 CTTTTTTT CTTTTTTTT 12
5 SRR2584863     CCGC    CCGCGC 10
6 SRR2584863        C         T 10
</code></pre></div></div>

<p>But our code isn’t very easy to understand and we’d need to see the output to make 
sure we had the correct columns.</p>

<p>Enter <code class="language-plaintext highlighter-rouge">dplyr</code>!</p>

<h4 id="how-to-select-columns">How to <strong><code class="language-plaintext highlighter-rouge">select()</code></strong> columns</h4>

<p>To select columns of a data frame, use <code class="language-plaintext highlighter-rouge">select()</code>. The first argument to this 
function is the data frame (<code class="language-plaintext highlighter-rouge">variants</code>), and the subsequent arguments are the 
columns to keep.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">select</span><span class="p">(</span><span class="n">variants</span><span class="p">,</span><span class="w"> </span><span class="n">sample_id</span><span class="p">,</span><span class="w"> </span><span class="n">REF</span><span class="p">,</span><span class="w"> </span><span class="n">ALT</span><span class="p">,</span><span class="w"> </span><span class="n">DP</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>   sample_id      REF       ALT DP
1 SRR2584863        T         G  4
2 SRR2584863        G         T  6
3 SRR2584863        G         T 10
4 SRR2584863 CTTTTTTT CTTTTTTTT 12
5 SRR2584863     CCGC    CCGCGC 10
6 SRR2584863        C         T 10
</code></pre></div></div>

<p>This code is much easier to understand!</p>

<p>To select all columns <em>except</em> certain ones, put a “-“ in front of
the variable to exclude it.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">select</span><span class="p">(</span><span class="n">variants</span><span class="p">,</span><span class="w"> </span><span class="o">-</span><span class="n">CHROM</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>   sample_id    POS ID      REF       ALT QUAL FILTER INDEL IDV IMF DP
1 SRR2584863   9972 NA        T         G   91     NA FALSE  NA  NA  4
2 SRR2584863 263235 NA        G         T   85     NA FALSE  NA  NA  6
3 SRR2584863 281923 NA        G         T  217     NA FALSE  NA  NA 10
4 SRR2584863 433359 NA CTTTTTTT CTTTTTTTT   64     NA  TRUE  12 1.0 12
5 SRR2584863 473901 NA     CCGC    CCGCGC  228     NA  TRUE   9 0.9 10
6 SRR2584863 648692 NA        C         T  210     NA FALSE  NA  NA 10
        VDB RPB MQB BQB     MQSB       SGB     MQ0F ICB HOB AC AN     DP4 MQ
1 0.0257451  NA  NA  NA       NA -0.556411 0.000000  NA  NA  1  1 0,0,0,4 60
2 0.0961330   1   1   1       NA -0.590765 0.166667  NA  NA  1  1 0,1,0,5 33
3 0.7740830  NA  NA  NA 0.974597 -0.662043 0.000000  NA  NA  1  1 0,0,4,5 60
4 0.4777040  NA  NA  NA 1.000000 -0.676189 0.000000  NA  NA  1  1 0,1,3,8 60
5 0.6595050  NA  NA  NA 0.916482 -0.662043 0.000000  NA  NA  1  1 1,0,2,7 60
6 0.2680140  NA  NA  NA 0.916482 -0.670168 0.000000  NA  NA  1  1 0,0,7,3 60
                                                               Indiv gt_PL
1 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 121,0
2 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 112,0
3 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 247,0
4 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam  91,0
5 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 255,0
6 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 240,0
  gt_GT gt_GT_alleles
1     1             G
2     1             T
3     1             T
4     1     CTTTTTTTT
5     1        CCGCGC
6     1             T
</code></pre></div></div>

<h4 id="tibbles-and-as_tibble">Tibbles and <strong><code class="language-plaintext highlighter-rouge">as_tibble()</code></strong></h4>

<p>One thing we still have is an issue with how our output prints - it takes up a 
lot of space in our console, and it’s a little confusing to read. We might get 
a warning like <code class="language-plaintext highlighter-rouge">[ reached 'max' / getOption("max.print") -- omitted 551 rows ]</code>
if we’re showing too many rows. We might usually just want to view the first few 
rows to check our output makes sense, not the first few hundred!</p>

<p>Luckily, the <code class="language-plaintext highlighter-rouge">tidyverse</code> has another package to help us: <code class="language-plaintext highlighter-rouge">tibble</code>.</p>

<p>Using the function <code class="language-plaintext highlighter-rouge">as_tibble()</code> will convert our <code class="language-plaintext highlighter-rouge">data.frame</code> object into the 
slightly friendlier <code class="language-plaintext highlighter-rouge">"tibble"</code> format, which is officially known as a <code class="language-plaintext highlighter-rouge">tbl_df</code>.</p>

<p>Tibbles have better formatting and printing and stricter handling of data types, 
and are the ‘central data structure’ for the <code class="language-plaintext highlighter-rouge">tidyverse</code> set of packages. 
They also help make our data manipulation much easier!</p>

<p>We use our variants data frame as the argument</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">variants</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as_tibble</span><span class="p">(</span><span class="n">variants</span><span class="p">)</span><span class="w">
</span><span class="n">select</span><span class="p">(</span><span class="n">variants</span><span class="p">,</span><span class="w"> </span><span class="n">sample_id</span><span class="p">,</span><span class="w"> </span><span class="n">REF</span><span class="p">,</span><span class="w"> </span><span class="n">ALT</span><span class="p">,</span><span class="w"> </span><span class="n">DP</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code># A tibble: 801 x 4
   sample_id  REF                    ALT                                      DP
   &lt;chr&gt;      &lt;chr&gt;                  &lt;chr&gt;                                 &lt;int&gt;
 1 SRR2584863 T                      G                                         4
 2 SRR2584863 G                      T                                         6
 3 SRR2584863 G                      T                                        10
 4 SRR2584863 CTTTTTTT               CTTTTTTTT                                12
 5 SRR2584863 CCGC                   CCGCGC                                   10
 6 SRR2584863 C                      T                                        10
 7 SRR2584863 C                      A                                         8
 8 SRR2584863 G                      A                                        11
 9 SRR2584863 ACAGCCAGCCAGCCAGCCAGC… ACAGCCAGCCAGCCAGCCAGCCAGCCAGCCAGCCAG…     3
10 SRR2584863 AT                     ATT                                       7
# … with 791 more rows
</code></pre></div></div>

<p>This looks much better!</p>

<h4 id="how-to-filter-rows">How to <strong><code class="language-plaintext highlighter-rouge">filter()</code></strong> rows</h4>

<p>To choose rows, use <code class="language-plaintext highlighter-rouge">filter()</code>:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">filter</span><span class="p">(</span><span class="n">variants</span><span class="p">,</span><span class="w"> </span><span class="n">sample_id</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"SRR2584863"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code># A tibble: 25 x 29
   sample_id CHROM    POS ID    REF   ALT    QUAL FILTER INDEL   IDV    IMF
   &lt;chr&gt;     &lt;chr&gt;  &lt;int&gt; &lt;lgl&gt; &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt; &lt;lgl&gt;  &lt;lgl&gt; &lt;int&gt;  &lt;dbl&gt;
 1 SRR25848… CP00… 9.97e3 NA    T     G        91 NA     FALSE    NA NA    
 2 SRR25848… CP00… 2.63e5 NA    G     T        85 NA     FALSE    NA NA    
 3 SRR25848… CP00… 2.82e5 NA    G     T       217 NA     FALSE    NA NA    
 4 SRR25848… CP00… 4.33e5 NA    CTTT… CTTT…    64 NA     TRUE     12  1    
 5 SRR25848… CP00… 4.74e5 NA    CCGC  CCGC…   228 NA     TRUE      9  0.9  
 6 SRR25848… CP00… 6.49e5 NA    C     T       210 NA     FALSE    NA NA    
 7 SRR25848… CP00… 1.33e6 NA    C     A       178 NA     FALSE    NA NA    
 8 SRR25848… CP00… 1.73e6 NA    G     A       225 NA     FALSE    NA NA    
 9 SRR25848… CP00… 2.10e6 NA    ACAG… ACAG…    56 NA     TRUE      2  0.667
10 SRR25848… CP00… 2.33e6 NA    AT    ATT     167 NA     TRUE      7  1    
# … with 15 more rows, and 18 more variables: DP &lt;int&gt;, VDB &lt;dbl&gt;, RPB &lt;dbl&gt;,
#   MQB &lt;dbl&gt;, BQB &lt;dbl&gt;, MQSB &lt;dbl&gt;, SGB &lt;dbl&gt;, MQ0F &lt;dbl&gt;, ICB &lt;lgl&gt;,
#   HOB &lt;lgl&gt;, AC &lt;int&gt;, AN &lt;int&gt;, DP4 &lt;chr&gt;, MQ &lt;int&gt;, Indiv &lt;chr&gt;,
#   gt_PL &lt;chr&gt;, gt_GT &lt;int&gt;, gt_GT_alleles &lt;chr&gt;
</code></pre></div></div>

<p>Note that this is equivalent to the base R code below, but is easier to read!</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">variants</span><span class="p">[</span><span class="n">variants</span><span class="o">$</span><span class="n">sample_id</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"SRR2584863"</span><span class="p">,]</span><span class="w">
</span></code></pre></div></div>

<h3 id="pipes">Pipes</h3>

<p>But what if you wanted to select <em>and</em> filter? We can do this with pipes.</p>

<p>Pipes let you take the output of one function and send it directly to the next, 
which is useful when you need to many things to the same data set. It was
possible to do this before pipes were added to R, but it was 
much messier and more difficult.</p>

<p>Pipes in R look like <code class="language-plaintext highlighter-rouge">%&gt;%</code> and are made available via the <code class="language-plaintext highlighter-rouge">magrittr</code> package, 
which is installed as part of <code class="language-plaintext highlighter-rouge">dplyr</code>. If you use RStudio, you can type the pipe 
with <kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>M</kbd> if you’re using a PC,
or <kbd>Cmd</kbd> + <kbd>Shift</kbd> + <kbd>M</kbd> if you’re using a Mac.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">variants</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">filter</span><span class="p">(</span><span class="n">sample_id</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"SRR2584863"</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">select</span><span class="p">(</span><span class="n">REF</span><span class="p">,</span><span class="w"> </span><span class="n">ALT</span><span class="p">,</span><span class="w"> </span><span class="n">DP</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code># A tibble: 25 x 3
   REF                        ALT                                             DP
   &lt;chr&gt;                      &lt;chr&gt;                                        &lt;int&gt;
 1 T                          G                                                4
 2 G                          T                                                6
 3 G                          T                                               10
 4 CTTTTTTT                   CTTTTTTTT                                       12
 5 CCGC                       CCGCGC                                          10
 6 C                          T                                               10
 7 C                          A                                                8
 8 G                          A                                               11
 9 ACAGCCAGCCAGCCAGCCAGCCAGC… ACAGCCAGCCAGCCAGCCAGCCAGCCAGCCAGCCAGCCAGCCA…     3
10 AT                         ATT                                              7
# … with 15 more rows
</code></pre></div></div>

<p>In the above code, we use the pipe to send the <code class="language-plaintext highlighter-rouge">variants</code> dataset first through
<code class="language-plaintext highlighter-rouge">filter()</code>, to keep rows where <code class="language-plaintext highlighter-rouge">sample_id</code> matches a particular sample, and then
through <code class="language-plaintext highlighter-rouge">select()</code> to keep only the <code class="language-plaintext highlighter-rouge">REF</code>, <code class="language-plaintext highlighter-rouge">ALT</code>, and <code class="language-plaintext highlighter-rouge">DP</code> columns. Since <code class="language-plaintext highlighter-rouge">%&gt;%</code> 
takes the object on its left and passes it as the first argument to the function 
on its right, we don’t need to explicitly include the data frame as an argument
to the <code class="language-plaintext highlighter-rouge">filter()</code> and <code class="language-plaintext highlighter-rouge">select()</code> functions any more.</p>

<p>Some may find it helpful to read the pipe like the word “then”. For instance,
in the above example, we took the data frame <code class="language-plaintext highlighter-rouge">variants</code>, <em>then</em> we <code class="language-plaintext highlighter-rouge">filter</code>ed
for rows where <code class="language-plaintext highlighter-rouge">sample_id</code> was SRR2584863, <em>then</em> we <code class="language-plaintext highlighter-rouge">select</code>ed the <code class="language-plaintext highlighter-rouge">REF</code>, <code class="language-plaintext highlighter-rouge">ALT</code>, 
and <code class="language-plaintext highlighter-rouge">DP</code> columns, <em>then</em> we showed only the first six rows.</p>

<p>The <strong><code class="language-plaintext highlighter-rouge">dplyr</code></strong> functions by themselves are somewhat simple,
but by combining them into linear workflows with the pipe, we can accomplish
more complex manipulations of data frames.</p>

<p>NOTE: Pipes work with non-<code class="language-plaintext highlighter-rouge">dplyr</code> functions, too, as long as the <code class="language-plaintext highlighter-rouge">dplyr</code> or <code class="language-plaintext highlighter-rouge">magrittr</code> package is loaded.</p>

<p>If we want to create a new object with this smaller version of the data we
can do so by assigning it a new name:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">SRR2584863_variants</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">variants</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">filter</span><span class="p">(</span><span class="n">sample_id</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"SRR2584863"</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">select</span><span class="p">(</span><span class="n">REF</span><span class="p">,</span><span class="w"> </span><span class="n">ALT</span><span class="p">,</span><span class="w"> </span><span class="n">DP</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p>This new object includes all of the data from this sample. Let’s look at just
the first six rows to confirm it’s what we want:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">head</span><span class="p">(</span><span class="n">SRR2584863_variants</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code># A tibble: 6 x 3
  REF      ALT          DP
  &lt;chr&gt;    &lt;chr&gt;     &lt;int&gt;
1 T        G             4
2 G        T             6
3 G        T            10
4 CTTTTTTT CTTTTTTTT    12
5 CCGC     CCGCGC       10
6 C        T            10
</code></pre></div></div>

<blockquote class="challenge">
  <h2 id="exercise-pipe-and-filter">Exercise: Pipe and filter</h2>

  <p>Starting with the <code class="language-plaintext highlighter-rouge">variants</code> dataframe, use pipes to subset the data
to include only observations from SRR2584863 sample,
where the filtered
depth (DP) is at least 10. Retain only the columns <code class="language-plaintext highlighter-rouge">REF</code>, <code class="language-plaintext highlighter-rouge">ALT</code>, 
and <code class="language-plaintext highlighter-rouge">POS</code>.</p>

  <blockquote class="solution">
    <h2 id="solution">Solution</h2>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="w"> </span><span class="n">variants</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">   
 </span><span class="n">filter</span><span class="p">(</span><span class="n">sample_id</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"SRR2584863"</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">DP</span><span class="w"> </span><span class="o">&gt;=</span><span class="w"> </span><span class="m">10</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
 </span><span class="n">select</span><span class="p">(</span><span class="n">REF</span><span class="p">,</span><span class="w"> </span><span class="n">ALT</span><span class="p">,</span><span class="w"> </span><span class="n">POS</span><span class="p">)</span><span class="w">
</span></code></pre></div>    </div>

    <div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code># A tibble: 16 x 3
   REF      ALT           POS
   &lt;chr&gt;    &lt;chr&gt;       &lt;int&gt;
 1 G        T          281923
 2 CTTTTTTT CTTTTTTTT  433359
 3 CCGC     CCGCGC     473901
 4 C        T          648692
 5 G        A         1733343
 6 A        C         2446984
 7 G        T         2618472
 8 A        T         2665639
 9 G        A         2999330
10 A        C         3339313
11 C        A         3401754
12 A        C         3488669
13 G        T         3909807
14 A        G         4100183
15 A        C         4201958
16 TGG      T         4431393
</code></pre></div>    </div>
  </blockquote>
</blockquote>

<h3 id="mutate">Mutate</h3>

<p>Frequently you’ll want to create new columns based on the values in existing
columns, for example to do unit conversions or find the ratio of values in two
columns. For this we’ll use the <code class="language-plaintext highlighter-rouge">dplyr</code> function <code class="language-plaintext highlighter-rouge">mutate()</code>.</p>

<p>We have a column titled “QUAL”. This is a Phred-scaled confidence
score that a polymorphism exists at this position given the sequencing
data. Lower QUAL scores indicate low probability of a polymorphism
existing at that site. We can convert the confidence value QUAL
to a probability value according to the formula:</p>

<p>Probability = 1- 10 ^ -(QUAL/10)</p>

<p>Let’s add a column (<code class="language-plaintext highlighter-rouge">POLPROB</code>) to our <code class="language-plaintext highlighter-rouge">variants</code> dataframe that shows 
the probability of a polymorphism at that site given the data.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">variants</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">mutate</span><span class="p">(</span><span class="n">POLPROB</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="p">(</span><span class="m">10</span><span class="w"> </span><span class="o">^</span><span class="w"> </span><span class="o">-</span><span class="p">(</span><span class="n">QUAL</span><span class="o">/</span><span class="m">10</span><span class="p">)))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">glimpse</span><span class="p">()</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Rows: 801
Columns: 30
$ sample_id     &lt;chr&gt; "SRR2584863", "SRR2584863", "SRR2584863", "SRR2584863",…
$ CHROM         &lt;chr&gt; "CP000819.1", "CP000819.1", "CP000819.1", "CP000819.1",…
$ POS           &lt;int&gt; 9972, 263235, 281923, 433359, 473901, 648692, 1331794, …
$ ID            &lt;lgl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA,…
$ REF           &lt;chr&gt; "T", "G", "G", "CTTTTTTT", "CCGC", "C", "C", "G", "ACAG…
$ ALT           &lt;chr&gt; "G", "T", "T", "CTTTTTTTT", "CCGCGC", "T", "A", "A", "A…
$ QUAL          &lt;dbl&gt; 91.0000, 85.0000, 217.0000, 64.0000, 228.0000, 210.0000…
$ FILTER        &lt;lgl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA,…
$ INDEL         &lt;lgl&gt; FALSE, FALSE, FALSE, TRUE, TRUE, FALSE, FALSE, FALSE, T…
$ IDV           &lt;int&gt; NA, NA, NA, 12, 9, NA, NA, NA, 2, 7, NA, NA, NA, NA, NA…
$ IMF           &lt;dbl&gt; NA, NA, NA, 1.000000, 0.900000, NA, NA, NA, 0.666667, 1…
$ DP            &lt;int&gt; 4, 6, 10, 12, 10, 10, 8, 11, 3, 7, 9, 20, 12, 19, 15, 1…
$ VDB           &lt;dbl&gt; 0.0257451, 0.0961330, 0.7740830, 0.4777040, 0.6595050, …
$ RPB           &lt;dbl&gt; NA, 1.000000, NA, NA, NA, NA, NA, NA, NA, NA, 0.900802,…
$ MQB           &lt;dbl&gt; NA, 1.0000000, NA, NA, NA, NA, NA, NA, NA, NA, 0.150134…
$ BQB           &lt;dbl&gt; NA, 1.000000, NA, NA, NA, NA, NA, NA, NA, NA, 0.750668,…
$ MQSB          &lt;dbl&gt; NA, NA, 0.974597, 1.000000, 0.916482, 0.916482, 0.90080…
$ SGB           &lt;dbl&gt; -0.556411, -0.590765, -0.662043, -0.676189, -0.662043, …
$ MQ0F          &lt;dbl&gt; 0.000000, 0.166667, 0.000000, 0.000000, 0.000000, 0.000…
$ ICB           &lt;lgl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA,…
$ HOB           &lt;lgl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA,…
$ AC            &lt;int&gt; 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1…
$ AN            &lt;int&gt; 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1…
$ DP4           &lt;chr&gt; "0,0,0,4", "0,1,0,5", "0,0,4,5", "0,1,3,8", "1,0,2,7", …
$ MQ            &lt;int&gt; 60, 33, 60, 60, 60, 60, 60, 60, 60, 60, 25, 60, 10, 60,…
$ Indiv         &lt;chr&gt; "/home/dcuser/dc_workshop/results/bam/SRR2584863.aligne…
$ gt_PL         &lt;chr&gt; "121,0", "112,0", "247,0", "91,0", "255,0", "240,0", "2…
$ gt_GT         &lt;int&gt; 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1…
$ gt_GT_alleles &lt;chr&gt; "G", "T", "T", "CTTTTTTTT", "CCGCGC", "T", "A", "A", "A…
$ POLPROB       &lt;dbl&gt; 1.0000000, 1.0000000, 1.0000000, 0.9999996, 1.0000000, …
</code></pre></div></div>

<p>NOTE: We’ve used the <code class="language-plaintext highlighter-rouge">glimpse()</code> function to show only the first few 
values of data for each variable - you can see it’s quite similar to 
<code class="language-plaintext highlighter-rouge">str()</code> in how it shows the output.  You could also use <code class="language-plaintext highlighter-rouge">head()</code> in the same way 
to just show 6 rows rather than the 10 rows shown by default with tibbles.</p>

<!-- also mention print() ? -->

<blockquote class="challenge">
  <h2 id="exercise">Exercise</h2>
  <p>There are a lot of columns in our dataset, so let’s just look at the
<code class="language-plaintext highlighter-rouge">sample_id</code>, <code class="language-plaintext highlighter-rouge">POS</code>, <code class="language-plaintext highlighter-rouge">QUAL</code>, and <code class="language-plaintext highlighter-rouge">POLPROB</code> columns for now. Add a 
line to the above code to only show those columns and show the output via <code class="language-plaintext highlighter-rouge">glimpse()</code>.</p>

  <blockquote class="solution">
    <h2 id="solution-1">Solution</h2>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">variants</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
</span><span class="n">mutate</span><span class="p">(</span><span class="n">POLPROB</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="m">10</span><span class="w"> </span><span class="o">^</span><span class="w"> </span><span class="o">-</span><span class="p">(</span><span class="n">QUAL</span><span class="o">/</span><span class="m">10</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
</span><span class="n">select</span><span class="p">(</span><span class="n">sample_id</span><span class="p">,</span><span class="w"> </span><span class="n">POS</span><span class="p">,</span><span class="w"> </span><span class="n">QUAL</span><span class="p">,</span><span class="w"> </span><span class="n">POLPROB</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
</span><span class="n">glimpse</span><span class="p">()</span><span class="w">
</span></code></pre></div>    </div>

    <div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Rows: 801
Columns: 4
$ sample_id &lt;chr&gt; "SRR2584863", "SRR2584863", "SRR2584863", "SRR2584863", "SR…
$ POS       &lt;int&gt; 9972, 263235, 281923, 433359, 473901, 648692, 1331794, 1733…
$ QUAL      &lt;dbl&gt; 91.0000, 85.0000, 217.0000, 64.0000, 228.0000, 210.0000, 17…
$ POLPROB   &lt;dbl&gt; 1.0000000, 1.0000000, 1.0000000, 0.9999996, 1.0000000, 1.00…
</code></pre></div>    </div>
  </blockquote>
</blockquote>

<h3 id="split-apply-combine-data-analysis-and-the-summarize-function">Split-apply-combine data analysis and the summarize() function</h3>

<p>Many data analysis tasks can be approached using the <em>“split-apply-combine”</em>
paradigm: split the data into groups, apply some analysis to each group, and
then combine the results. <code class="language-plaintext highlighter-rouge">dplyr</code> makes this very easy through the use of the
<code class="language-plaintext highlighter-rouge">group_by()</code> function, which splits the data into groups.</p>

<h4 id="the-summarize-function">The <code class="language-plaintext highlighter-rouge">summarize()</code> function</h4>

<p><code class="language-plaintext highlighter-rouge">group_by()</code> is often used together with <code class="language-plaintext highlighter-rouge">summarize()</code>, which collapses each
group into a single-row summary of that group.  <code class="language-plaintext highlighter-rouge">group_by()</code> takes as arguments
the column names that contain the <strong>categorical</strong> variables for which you want
to calculate the summary statistics.</p>

<p>For example, if we wanted to group by <code class="language-plaintext highlighter-rouge">sample_id</code> and find the number of rows 
of data for each sample, we would do:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">variants</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">group_by</span><span class="p">(</span><span class="n">sample_id</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">summarize</span><span class="p">(</span><span class="n">n_observations</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">n</span><span class="p">())</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>`summarise()` ungrouping output (override with `.groups` argument)
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code># A tibble: 3 x 2
  sample_id  n_observations
  &lt;chr&gt;               &lt;int&gt;
1 SRR2584863             25
2 SRR2584866            766
3 SRR2589044             10
</code></pre></div></div>

<p>Here the summary function used was <code class="language-plaintext highlighter-rouge">n()</code> to find the count for each
group, which we displayed in a column which we called <code class="language-plaintext highlighter-rouge">n_observations</code>.</p>

<p>We can also apply many other functions  to individual columns
to get other summary statistics. For example, we can use built-in functions like
<code class="language-plaintext highlighter-rouge">mean()</code>, <code class="language-plaintext highlighter-rouge">median()</code>, <code class="language-plaintext highlighter-rouge">min()</code>, and <code class="language-plaintext highlighter-rouge">max()</code>. These are called “built-in 
functions” because they come with R and don’t require that you install any 
additional packages.</p>

<p>So to view the highest filtered depth (<code class="language-plaintext highlighter-rouge">DP</code>) for each sample:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">variants</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">group_by</span><span class="p">(</span><span class="n">sample_id</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">summarize</span><span class="p">(</span><span class="nf">max</span><span class="p">(</span><span class="n">DP</span><span class="p">))</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>`summarise()` ungrouping output (override with `.groups` argument)
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code># A tibble: 3 x 2
  sample_id  `max(DP)`
  &lt;chr&gt;          &lt;int&gt;
1 SRR2584863        20
2 SRR2584866        79
3 SRR2589044        16
</code></pre></div></div>

<!-- TODO: summarize() more than one variable (e.g. get min, mean sampling depth per sample) -->
<!-- TODO: summarize() more than one variable (e.g. get min, mean sampling depth per sample) -->
<!-- TODO: `dplyr::count()` as alternative to `n()` -->
<!-- TODO: `dplyr::arrange()` -->

<h3 id="handling-missing-values-in-data">Handling Missing Values In Data</h3>

<p>By default, all <strong>R functions operating on vectors that contains missing data 
will return NA</strong>.</p>

<p>It’s a way to make sure that users know they have missing
data, and make a conscious decision on how to deal with it. When
dealing with simple statistics like the mean, the easiest way to
ignore <code class="language-plaintext highlighter-rouge">NA</code> (the missing data) is to use <code class="language-plaintext highlighter-rouge">na.rm = TRUE</code> (<code class="language-plaintext highlighter-rouge">rm</code> stands for
remove).</p>

<p>When working with data frames (and tibbles!), we have slightly more options.</p>

<!-- TODO: add information about those options! -->

<!-- Add exercise about filtering out NA values either in filter() section, or -->
<!-- in pipe or mutate() section, with `filter(!is.na())` base-r-ish version or -->
<!-- tidyr::drop_na() which I think is more recent and 'tidyverse' canon version -->

<!-- ### Reshaping with gather and spread -->
<!-- TODO: Reshaping with gather and spread -->

<!-- You can group by multiple columns too. For example, we might want to  -->
<!-- know if ...  -->

<!-- ```{r, purl = FALSE} -->

<!-- ``` -->

<!-- You can also summarize multiple variables at the same time. Let's -->
<!-- count how many samples we have in each group using the `n()` function: -->

<!-- ```{r, purl = FALSE} -->

<!-- ``` -->

<h2 id="exporting-data">Exporting data</h2>

<p>Now that you have learned how to use <strong><code class="language-plaintext highlighter-rouge">dplyr</code></strong> to extract information from
or summarize your raw data, you may want to export these new data sets to share
them with your collaborators or for archival.</p>

<p>Similar to the <code class="language-plaintext highlighter-rouge">read.csv()</code> function used for reading CSV files into R, there is
a <code class="language-plaintext highlighter-rouge">write.csv()</code> function that generates CSV files from data frames.</p>

<p>Before using <code class="language-plaintext highlighter-rouge">write.csv()</code>, we should preferrably create a new folder, <code class="language-plaintext highlighter-rouge">data</code>,
in our working directory that will store this generated dataset. We don’t want
to write generated datasets in the same directory as our raw data. It’s good
practice to keep them separate.</p>

<p>Ideally, we should have a <code class="language-plaintext highlighter-rouge">data_raw</code> folder should only contain the raw, unaltered data, and should be left alone to make sure we don’t delete or modify it. In contrast, 
our script will generate the contents of the <code class="language-plaintext highlighter-rouge">data</code> directory, so even if the 
files it contains are deleted, we can always re-generate them.</p>

<p>In preparation for our next lesson on plotting, we are going to prepare a
cleaned up version of the data set that doesn’t include any missing data.</p>

<!-- TODO: export the dataset - make cleaned up version of dataset -->
<!-- * remove missing data -->
<!-- * filter to particular quality of data (over sequencing depth threshold?) -->
<!-- * KEEP IN MIND WHY? WHAT WILL WE PLOT WITH THIS LATER??? -->
<!-- * check dim() to ensure everyone's got same dataset -->
<!-- * change `eval=FALSE` to eval=TRUE for 1st and 3rd of these code blocks -->
<!-- * remove `echo=FALSE from first two of these code blocks -->

<p><a href="https://github.com/rstudio/cheatsheets/raw/master/data-transformation.pdf">Handy dplyr cheatsheet</a></p>

<p><em>Much of this lesson was copied or adapted from Jeff Hollister’s <a href="http://usepa.github.io/introR/2015/01/14/03-Clean/">materials</a></em></p>





<blockquote class="keypoints">
  <h2>Key Points</h2>
  <ul>
    
    <li><p>Use the <code class="language-plaintext highlighter-rouge">dplyr</code> package to manipulate dataframes.</p>
</li>
    
    <li><p>Use <code class="language-plaintext highlighter-rouge">select()</code> to choose variables from a dataframe.</p>
</li>
    
    <li><p>Use <code class="language-plaintext highlighter-rouge">filter()</code> to choose data based on values.</p>
</li>
    
    <li><p>Use <code class="language-plaintext highlighter-rouge">group_by()</code> and <code class="language-plaintext highlighter-rouge">summarize()</code> to work with subsets of data.</p>
</li>
    
    <li><p>Use <code class="language-plaintext highlighter-rouge">mutate()</code> to create new variables.</p>
</li>
    
  </ul>
</blockquote>

</article>



























  
  











<div class="row">
  <div class="col-xs-1">
    <h3 class="text-left">
      
      <a href="../03-basics-factors-dataframes/index.html"><span class="glyphicon glyphicon-menu-left" aria-hidden="true"></span><span class="sr-only">previous episode</span></a>
      
    </h3>
  </div>
  <div class="col-xs-10">
    
  </div>
  <div class="col-xs-1">
    <h3 class="text-right">
      
      <a href="../05-data-visualization/index.html"><span class="glyphicon glyphicon-menu-right" aria-hidden="true"></span><span class="sr-only">next episode</span></a>
      
    </h3>
  </div>
</div>



      
      






<footer>
  <hr/>
  <div class="row">
    <div class="col-md-6 license" id="license-info" align="left">
	
	Licensed under <a href="https://creativecommons.org/licenses/by/4.0/">CC-BY 4.0</a> 2018–2020
	by <a href="https://carpentries.org/">The Carpentries</a>
        <br>
        Licensed under <a href="https://creativecommons.org/licenses/by/4.0/">CC-BY 4.0</a> 2016–2018
	by <a href="http://datacarpentry.org">Data Carpentry</a>
	
    </div>
    <div class="col-md-6 help-links" align="right">
	
	
	<a href="/edit//_episodes_rmd/04-dplyr.Rmd" data-checker-ignore>Edit on GitHub</a>
	
	
	/
	<a href="/blob//CONTRIBUTING.md" data-checker-ignore>Contributing</a>
	/
	<a href="/">Source</a>
	/
	<a href="/blob//CITATION" data-checker-ignore>Cite</a>
	/
	<a href="mailto:team@carpentries.org">Contact</a>
    </div>
  </div>
  <p class="text-muted text-right">
    <small><i>Using <a href="https://github.com/carpentries/carpentries-theme/">The Carpentries theme</a> &mdash; Site last built on: 2020-12-06 21:20:29 +0000.</i></small>
  </p>
</footer>

      
    </div>
    
<script src="../assets/js/jquery.min.js"></script>
<script src="../assets/js/bootstrap.min.js"></script>
<script src="../assets/js/lesson.js"></script>
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');
  ga('create', 'UA-37305346-2', 'auto');
  ga('send', 'pageview');
</script>

  </body>
</html>
